samples/apps/tinyRA/tinyra/tui.py

```diff
@@ -0,0 +1,1186 @@
+import os
+import asyncio
+import ast
+import configparser
+import platform
+import json
+import logging
+import argparse
+import shutil
+import sqlite3
+import aiosqlite
+
+from typing import List, Dict
+
+from textual import work
+from textual.app import App, ComposeResult
+from textual.screen import Screen, ModalScreen
+from textual.containers import ScrollableContainer, Grid, Container
+from textual.widget import Widget
+from textual.widgets import Pretty
+from textual.widgets import Footer, Header, Markdown, Static, Input, DirectoryTree, Label
+from textual.widgets import Button, TabbedContent, ListView, ListItem
+from textual.widgets import TextArea
+from textual.reactive import reactive
+from textual.message import Message
+from textual.events import Key
+
+from autogen import config_list_from_json
+from autogen import Agent, AssistantAgent, UserProxyAgent
+
+
+class Tool:
+    def __init__(self, name: str, code: str, description: str = None, id: int = None):
+        self.id = id
+        self.name = name
+        self.code = code
+
+        if not description and len(code) > 0:
+            # Parse the function string into an AST
+            module = ast.parse(code)
+
+            # Get the function definition from the AST
+            function_def = module.body[0]
+
+            # Extract the docstring
+            self.description = ast.get_docstring(function_def)
+        else:
+            self.description = description or self.name
+
+
+def string_to_function(code: str):
+    function_name = code.split("def ")[1].split("(")[0]
+    global_namespace = globals()
+    local_namespace = {}
+    exec(code, global_namespace, local_namespace)
```

Comment:
@jackgerrits I need your help with this one. 

Input: a string containing python code (e.g., a sync python function)
Intended behavior: register the python function with llm and with code executor

The reason I wrote this function is because I need to go from string -> instance of the function, which I could register with the agents.